### PR TITLE
Implemented FirewallPluginInterface

### PR DESCRIPTION
**Pull request by aschempp**
see #24

This solves the issue that security firewall configuration cannot be extended (`cannotBeOverwritten` is set on the `TreeBuilder`). I wonder if it would make sense to implement this as a general solution as we don't know if there are other such nodes?

### CODE REVIEW

**Review by Toflar — CHANGES_REQUESTED** [review]
Tests fail. Other than that looks good, thank you!

> I wonder if it would make sense to implement this as a general solution as we don't know if there are other such nodes?

To be honest, I can't think of one at the moment and we should not overcomplicate things right now as we're still in the early stages. I think the firewall is the 99% case and if there will be another one we might still build a general solution. Even if we build the general solution I still think it is a good thing to keep the `FirewallPluginInterface` because it is more specific in naming and pretty obvious why it's there. It might still map to a new, more general solution, in the future so we don't block anything here imho :)

File: src/ContaoManager/Plugin.php [hunk 1]
```diff
@@ -84,6 +86,17 @@ public function registerContainerConfiguration(LoaderInterface $loader, array $m
     }
 
     /**
+     * {@inheritdoc)
+     */
+    public function getFirewallConfig(array $config)
+    {
+        return array_merge(
+            $config,
+            Yaml::parse(file_get_contents(__DIR__.'/../Resources/contao-manager/security_firewalls.yml'))
```
Comment (leofeyer) [inline src/ContaoManager/Plugin.php]: Why are you not passing the loader as argument so we can do

```php
$loader->load('@ContaoManagerBundle/Resources/contao-manager/security_firewalls.yml');
```

?

Comment (Toflar) [reply src/ContaoManager/Plugin.php]: It has to be an array so you can order it. That's the whole point about this plugin :)

**Review by leofeyer — APPROVED** [review]

**Comment by aschempp** [comment]
I have rewritten the implementation to use a more generic approach. The new `ExtensionPluginInterface` allows to process all extension configurations before they are used (basically, before they are processed by the extension).

This has several advantages:
 1. It's not specific to firewalls
 2. In regular cases, the interface is not needed at all
 3. It allows a plugin to modify **all** extension configuration from all bundles (even prepend), plugins and the application configuration itself.

The tests are currently failing because of #29

File: src/HttpKernel/ContaoKernel.php [hunk 2]
```diff
@@ -172,6 +174,21 @@ public function registerContainerConfiguration(LoaderInterface $loader)
     /**
      * {@inheritdoc}
      */
+    protected function getContainerBuilder()
+    {
+        $container = new PluginContainerBuilder($this->getPluginLoader(), []);
+        $container->getParameterBag()->add($this->getKernelParameters());
+
+        if (class_exists('ProxyManager\Configuration') && class_exists('Symfony\Bridge\ProxyManager\LazyProxy\Instantiator\RuntimeInstantiator')) {
```
Comment (leofeyer) [inline src/HttpKernel/ContaoKernel.php:182]: You could use `Configuration::class` and `RuntimeInstantiator::class` here.

Comment (aschempp) [reply src/HttpKernel/ContaoKernel.php:182]: That code is a direct copy from Symfony so I'd rather not change it…?

**Review by Toflar — APPROVED** [review]

File: src/HttpKernel/ContaoKernel.php (hunk 2 above)
Comment (leofeyer) [reply src/HttpKernel/ContaoKernel.php:182]: Agreed. :)

**Pull request merged**